exercises/tree-building/src/test/java/BuildTreeTest.java

```diff
@@ -0,0 +1,197 @@
+        records.add(new Record(5, 1));
+        records.add(new Record(6, 2));
+
+        TreeNode root = new BuildTree().buildTree(records);
+        assertNodeIsBranch(root, 0, 2);
+        assertNodeIsBranch(root.children.get(0), 1, 3);
+        assertNodeIsBranch(root.children.get(1), 2, 1);
+        assertNodeIsLeaf(root.children.get(0).children.get(0), 3);
+        assertNodeIsLeaf(root.children.get(0).children.get(1), 4);
+        assertNodeIsLeaf(root.children.get(0).children.get(2), 5);
+        assertNodeIsLeaf(root.children.get(1).children.get(0), 6);
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void testRootNodeHasParent() throws InvalidRecordsException {
+        expectedException.expect(InvalidRecordsException.class);
+        expectedException.expectMessage("Invalid Records");
+
+        ArrayList<Record> records = new ArrayList<>();
+        records.add(new Record(0, 1));
+        records.add(new Record(1, 0));
+
+        BuildTree test = new BuildTree();
+        test.buildTree(records);
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void testNoRootNode() throws InvalidRecordsException {
+        expectedException.expect(InvalidRecordsException.class);
+        expectedException.expectMessage("Invalid Records");
+
+        ArrayList<Record> records = new ArrayList<>();
+        records.add(new Record(1, 0));
+        records.add(new Record(2, 0));
+
+        BuildTree test = new BuildTree();
+        test.buildTree(records);
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void testNonContinuousRecords() throws InvalidRecordsException {
+        expectedException.expect(InvalidRecordsException.class);
+        expectedException.expectMessage("Invalid Records");
+
+        ArrayList<Record> records = new ArrayList<>();
+        records.add(new Record(2, 0));
+        records.add(new Record(4, 2));
+        records.add(new Record(1, 0));
+        records.add(new Record(0, 0));
+
+        BuildTree test = new BuildTree();
+        test.buildTree(records);
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void testCycleInDirectly() throws InvalidRecordsException {
+        expectedException.expect(InvalidRecordsException.class);
+        expectedException.expectMessage("Invalid Records");
+
+        ArrayList<Record> records = new ArrayList<>();
+        records.add(new Record(5, 2));
+        records.add(new Record(3, 2));
+        records.add(new Record(2, 6));
+        records.add(new Record(4, 1));
+        records.add(new Record(1, 0));
+        records.add(new Record(0, 0));
+        records.add(new Record(6, 3));
+
+        BuildTree test = new BuildTree();
+        test.buildTree(records);
+    }
+
+    private void assertNodeIsLeaf(TreeNode node, int nodeId) {
+        assertEquals(node.getNodeId(), nodeId);
+        assertEquals(node.children.size(), 0);
+    }
+
+    private void assertNodeIsBranch(TreeNode node, int nodeId, int childrenCount) {
```

Comment:
Same for this, I would move the node id check out of this method. And I would rename it to something like `assertNodeIsBranchWithNNumberOfChildren`. Otherwise it's not obvious that it also checks the number of children 🙂